libkbfs/file_data.go

```diff
@@ -585,52 +863,66 @@ func (fd *fileData) write(ctx context.Context, data []byte, off int64,
 		nCopied += fd.bsplit.CopyUntilSplit(
 			block, nextBlockOff < 0, data[nCopied:max], off+nCopied-startOff)
 
-		// TODO: support multiple levels of indirection.  Right now the
-		// code only does one but it should be straightforward to
-		// generalize, just annoying
-
-		// if we need another block but there are no more, then make one
+		// If we need another block but there are no more, then make one.
 		switchToIndirect := false
-		if nCopied < n && nextBlockOff < 0 {
-			// If the block doesn't already have a parent block, make one.
-			if ptr == fd.file.tailPointer() {
-				topBlock, err = fd.createIndirectBlock(
-					df, DefaultNewBlockDataVersion(false))
+		if nCopied < n {
+			var newTopBlock *FileBlock
+			needExtendFile := nextBlockOff < 0
+			needFillHole := off+nCopied < nextBlockOff
+			nextOff := startOff + int64(len(block.Contents))
+			if nextOff < off+nCopied {
+				// We are writing somewhere inside a hole, not right
+				// at the start of it.
+				nextOff = off + nCopied
+			}
+			if needExtendFile || needFillHole {
+				// Make a new right block and update the parent's
+				// indirect block list, adding a level of indirection
+				// if needed.  If we're just filling a hole, the block
+				// will end up all the way to the right of the range,
+				// and its offset will be smaller than the block to
+				// its left -- we'll fix that up below.
+				var newDirtyPtrs []BlockPointer
+				fd.log.CDebugf(ctx, "Making new right block at nCopied=%d, nextOff=%d", nCopied, nextOff)
+				newTopBlock, newDirtyPtrs, err = fd.newRightBlock(
+					ctx, parentBlocks, nextOff, df,
+					DefaultNewBlockDataVersion(false))
 				if err != nil {
 					return newDe, nil, unrefs, newlyDirtiedChildBytes, 0, err
 				}
-				ptr = topBlock.IPtrs[0].BlockPointer
-				// The whole block needs to be re-uploaded as an
-				// indirect block, so track those dirty bytes and
-				// cache the block as dirty.
-				switchToIndirect = true
-			}
-
-			// Make a new right block and update the parent's
-			// indirect block list
-			err = fd.newRightBlock(ctx, fd.file.tailPointer(), topBlock,
-				startOff+int64(len(block.Contents)))
-			if err != nil {
-				return newDe, nil, unrefs, newlyDirtiedChildBytes, 0, err
-			}
-		} else if nCopied < n && off+nCopied < nextBlockOff {
-			// We need a new block to be inserted here
-			err = fd.newRightBlock(ctx, fd.file.tailPointer(), topBlock,
-				startOff+int64(len(block.Contents)))
-			if err != nil {
-				return newDe, nil, unrefs, newlyDirtiedChildBytes, 0, err
+				for _, p := range newDirtyPtrs {
+					dirtyMap[p] = true
+				}
+				if newTopBlock != nil {
+					if !topBlock.IsInd {
+						// The whole block needs to be re-uploaded as
+						// a child block with a new block pointer, so
+						// below we'll need to track those dirty bytes
+						// and cache the block as dirty.
```

Comment:
Yes, but the dirty bcache only tracks direct block bytes right now, not the bytes of indirect blocks, so in the case you mention nothing additional needs to be tracked.

Now that we can have more indirect blocks, we might think about changing that policy, but as of now indirect block bytes don't need tracking.